### PR TITLE
initial commit v2.0.0

### PR DESCRIPTION
**Pull request by cs-harsh-sharma**
### Descriptions:
PMDB: 29697
Mantis: 0982574
- Connector was broken, fixed the connector code to make it work.
- Fixed the 'JSON payload sent in the requests'.
- Added error messages and pagination support where ever possible.
- Updated playbooks and docs.
- Updated connector description, connector category and action input parameter descriptions. Also added tooltips where ever required.
- Added the following actions and playbooks: 
    - Search Reports
    - Download Report
    - Search Schedule 
    - Get Schedule Details
    - Search Option Profiles
    - Search Tags
    - Create Tag
    - Update Tag
    - Delete Tag
    - Search Users
- Updated the following action names:
    - Retrieve Scan Results > Get Scan Results
    - Retrieve Scan Status > Get Scan Status
    - Count Web Applications > Get Web Applications Count
- In action "Get Scan Count":
  - The type of input parameter "Scan ID", "Web App ID", "Web App Tag ID" is changed to text from integer.
  - The type of input parameter "Type", "Mode", "Status", "Authentication Status", "Result Status" is changed to multiselect from select.
  - The options "CANCELING", "ERROR", "CANCELED_WITH_RESULTS" are added and "CANCELING ERROR" is removed from the input parameter "Result Status".

- In action "Search Scans":
  - Renamed the input parameter "Web App Rag ID" to "Web App Tag ID".
  - The type of input parameter "Scan ID", "Web App ID", "Web App Tag ID" is changed to text from integer.
  - The type of input parameter "Type", "Mode", "Status", "Authentication Status", "Result Status" is changed to multiselect from select.
  - The options "CANCELING", "ERROR", "CANCELED_WITH_RESULTS" are added and "CANCELING ERROR" is removed from the input parameter "Result Status".
  - Input parameters "Limit" and "Offset" are added.

- In action "Launch Scans":
  - The type of input parameter "Scan Type", "Target Scanner Appliance Type" is changed to multiselect from select.

- In action "Get Web Applications Count":
  - The type of input parameter "ID", is changed to text from integer.
  - Input parameters "Tags ID", "Tags Name", "Created Date", "Updated Date", "Last Scan Status", "Last Scan Date" are added.

- In action "Search Web Applications":
  - Renamed the input parameter "Scan ID" to "ID", "Scan Name" to "Name".
  - The type of input parameter "ID", "Tags ID", "Tags Name" is changed to text from integer.
  - Input parameters "Scheduled", "Scanned", "Last Scan Status", "Last Scan Date", "Limit", "Offset", "Verbose" are added.

- In action "Get Web Application Details":
  - Renamed the input parameter "Scan ID" to "App ID".

- In action "Delete Web Applications":
  - Updated output schema.
#### UTCs:
- [x] Connector installation verified.
- [x] Connector logo verified.
- [x] Actions and Playbook list verified.
- [x] Playbook tags verified.
- [x] Input parameter tooltips verified.
- [x] All Playbook are in info mode verified.
- [x] All Playbook are in inactive mode verified.
#### Affected Actions:
 - [x] All

### CODE REVIEW

File: qualys-web-application-scanner/connector.py [hunk 1]
```diff
@@ -1,5 +1,10 @@
-from connectors.core.connector import Connector, get_logger, ConnectorError
+""" Copyright start
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/connector.py]: Change it to """
Copyright start
MIT License
Copyright (c) 2024 Fortinet Inc
Copyright end
"""

File: qualys-web-application-scanner/operations.py [hunk 2]
```diff
@@ -1,15 +1,26 @@
-import base64
+""" Copyright start
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/operations.py]: Change it to """
Copyright start
MIT License
Copyright (c) 2024 Fortinet Inc
Copyright end
"""

File: qualys-web-application-scanner/operations.py [hunk 3]
```diff
@@ -39,12 +55,18 @@ def make_rest_call(self, endpoint, params=None,
                                         params=params)
             if response.ok:
                 if response.status_code == 200:
-                    return response.json()
+                    if is_file_action:
+                        return response.content
+                    else:
+                        return response.json()
             # return json.loads(response.content.decode('utf-8')) if response.content.decode('utf-8') else ''
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/operations.py:63]: Remove commented code

File: qualys-web-application-scanner/playbooks/playbooks.json [hunk 4]
```diff
@@ -59,22 +59,19 @@
...
+              "uuid": "f1f88c30-c0c4-4245-804f-cf946a83d1e6",
               "@type": "WorkflowStep",
-              "name": "Scan Count",
+              "name": "Get Scan Count",
               "description": null,
               "status": null,
               "arguments": {
-                "name": "Qualys Web Scan",
+                "name": "Qualys Web Application Scanning(WAS)",
                 "config": "''",
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/playbooks/playbooks.json]: Change it to "config": "''" to "config": ""

File: qualys-web-application-scanner/info.json [hunk 5]
```diff
@@ -177,31 +184,31 @@
...
           "title": "Authentication Status",
           "required": false,
           "editable": true,
           "visible": true,
-          "type": "select",
-          "option": [
+          "type": "multiselect",
+          "options": [
             "NONE",
             "NOT_USED",
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/info.json:197]: Change the option in the correct format like NOT_USED to Not Used in all

File: qualys-web-application-scanner/info.json [hunk 6]
```diff
@@ -306,25 +317,28 @@
           "visible": true,
           "type": "text",
           "name": "webApp.name",
-          "description": "The name of the web application being scanned."
+          "description": "Specify the name of the web application being scanned.",
+          "tooltip": "Specify the name of the web application being scanned."
         },
         {
           "title": "Web App id",
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/info.json]: Change the title Web App ID

File: qualys-web-application-scanner/info.json [hunk 7]
```diff
@@ -342,77 +357,79 @@
...
+          "tooltip": "Specify the date and time to return all the scan which was launched after specified date and time in UTC date/time format (YYYY-MM-DDTHH:MM:SSZ)"
         },
         {
           "title": "Type",
           "required": false,
           "editable": true,
           "visible": true,
-          "type": "select",
-          "option": [
+          "type": "multiselect",
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/info.json:368]: Mention the type change in release notes.

File: qualys-web-application-scanner/info.json [hunk 8]
```diff
@@ -342,77 +357,79 @@
...
+          "description": "Specify the scan type. You can choose from the following options: VULNERABILITY or DISCOVERY."
         },
         {
           "title": "Mode",
           "required": false,
           "editable": true,
           "visible": true,
-          "type": "select",
-          "option": [
+          "type": "multiselect",
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/info.json:381]: Mention the type change in release notes. And do the same for others para as well.

File: qualys-web-application-scanner/info.json [hunk 9]
```diff
@@ -832,9 +879,10 @@
           "required": false,
           "editable": true,
           "visible": true,
-          "type": "integer",
+          "type": "text",
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/info.json:882]: Here as well for type.

File: qualys-web-application-scanner/info.json [hunk 10]
```diff
@@ -1217,7 +1431,980 @@
...
+          "required": false,
+          "editable": true,
+          "visible": true,
+          "type": "datetime",
+          "name": "lastScan.launchedDate",
+          "description": "Specify the date and time to return all the web application which was last scanned in WAS after specified date and time in UTC date/time format.",
+          "tooltip": "Specify the date and time to return all the web application which was last scanned in WAS after specified date and time in UTC date/time format."
+        },
+        {
+          "title": "Multi",
```
Comment (cs-utkarsh-surana) [inline qualys-web-application-scanner/info.json:1877]: Can you change the title to something more understandable

**Review by cs-utkarsh-surana — CHANGES_REQUESTED** [review]

File: qualys-web-application-scanner/connector.py (hunk 1 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/connector.py]: updated

File: qualys-web-application-scanner/playbooks/playbooks.json (hunk 4 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/playbooks/playbooks.json]: done

File: qualys-web-application-scanner/info.json (hunk 6 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/info.json]: done

File: qualys-web-application-scanner/info.json (hunk 7 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/info.json:368]: done

File: qualys-web-application-scanner/info.json (hunk 8 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/info.json:381]: done

File: qualys-web-application-scanner/info.json (hunk 9 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/info.json:882]: done

File: qualys-web-application-scanner/info.json (hunk 10 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/info.json:1877]: The API Doc uses same name and tooltip is provided with the input parameter to make it clear.

File: qualys-web-application-scanner/operations.py (hunk 3 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/operations.py:63]: Lets keep it as this is uncertified connector if something breaks may have to revert back.

File: qualys-web-application-scanner/info.json (hunk 5 above)
Comment (cs-harsh-sharma) [reply qualys-web-application-scanner/info.json:197]: This is a new version the options were added in the first release of connector. moreover, the customer is already used to the options.

**Pull request merged**